### PR TITLE
Remove LibSource

### PR DESCRIPTION
**Pull request by bjorn3**
The information is stored in used_crate_source too anyway.

Split out of #86105

r? @petrochenkov

### CODE REVIEW

**Comment by bjorn3** [comment]
Up to 0.2% regression with a few small improvements, so the other commit in #86105 was probably the cause of the large regression.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -256,22 +256,22 @@ pub fn each_linked_rlib(
...
-            LibSource::MetadataOnly => {
-                return Err(format!(
-                    "could not find rlib for: `{}`, found rmeta (metadata) file",
-                    name
-                ));
-            }
-            LibSource::None => return Err(format!("could not find rlib for: `{}`", name)),
+        let path = if let Some((path, _)) = &info.used_crate_source[&cnum].rlib {
+            path
+        } else if info.used_crate_source[&cnum].rmeta.is_some() {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:268]: The `info.used_crate_source[&cnum]` search can be done once instead of twice here.

**Comment by petrochenkov** [comment]
r=me with https://github.com/rust-lang/rust/pull/86877#discussion_r663936736 addressed.

**Pull request merged**